File: lintcode/word-ladder.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/word-ladder/

import string

class Solution:
    # @param start, a string
    # @param end, a string
    # @param dict, a set of string
    # @return an integer
    def ladderLength(self, start, end, dict):
        # write your code here
        alphabet = string.ascii_lowercase
        dict.add(end)
        queue = []
        queue.append((start, 1))
        while len(queue) > 0:
            word, dist = queue.pop(0)
            if word == end:
                return dist

            for j in alphabet:
                for i in xrange(len(word)):
                    if j != word[i]:
                        new_word = word[:i] + j + word[i+1:]
                        if new_word in dict:
                            dict.remove(new_word)
                            queue.append((new_word, dist + 1))
        return 0

if __name__ == '__main__':
    s = Solution()
    print s.ladderLength('hit', 'cog', set(["hot","dot","dog","lot","log"]))

File: leetcode/word-ladder.cc
// Question: https://leetcode.com/problems/word-ladder/

#include "headers.h"

class Solution {
public:
    int ladderLength(string beginWord, string endWord, unordered_set<string>& wordList) {
        wordList.insert(endWord);
        string alphabet = "abcdefghijklmnopqrstuvwxyz";
        queue<pair<string, int>> q;
        q.push(make_pair(beginWord, 1));
        while (!q.empty()) {
            pair<string, int> step = q.front();
            q.pop();
            if (step.first == endWord) {
                return step.second;
            }
            for (int i = 0; i < step.first.size(); ++i) {
                for (int j = 0; j < alphabet.size(); ++j) {
                    if (step.first.at(i) != alphabet.at(j)) {
                        string t = step.first.substr(0, i) + alphabet.substr(j, 1) + step.first.substr(i+1);
                        if (wordList.count(t) > 0) {
                            q.push(make_pair(t, step.second+1));
                            wordList.erase(wordList.find(t));
                        }
                    }
                }
            }
        }
        return 0;
    }
};

int main(int argc, char** argv) {
    unordered_set<string> words;
    words.insert("hot");
    words.insert("dot");
    words.insert("lot");
    words.insert("dog");
    words.insert("log");
    Solution s;
    cout << s.ladderLength("hit", "cog", words) << endl;
    return 0;
}
